### PR TITLE
Added batch info for arb receipt

### PR DESCRIPTION
**Pull request by yahgwai**
Also did a bit of organising.

Have the following open questions:
1. How to choose a delta in `getBatch` - and are we setting start/end block correctly?
2. Are any of the following fields nullabe?:
```
// l1SequenceNumber: bigNumber,
// parentRequestId: hash,
// indexInParent: number,
// arbType: number,
// arbSubType: number,
// l1BlockNumber: number,
// returnData: data, - could be empty
// returnCode: number, - never empty
// feeStats: {
//     prices: {
//         l1Transaction: bigNumber,
//         l1Calldata: bigNumber,
//         l2Storage: bigNumber,
//         l2Computation: bigNumber,
//     }
//     unitsUsed: {
//         l1Transaction: bigNumber,
//         l1Calldata: bigNumber,
//         l2Storage: bigNumber,
//         l2Computation: bigNumber,
//     }
//     paid: {
//         l1Transaction: bigNumber,
//         l1Calldata: bigNumber,
//         l2Storage: bigNumber,
//         l2Computation: bigNumber,
//     }
// }
```
3. The following dont have docs, and I couldnt find much about them. What do they mean?
```
// parentRequestId: string
//   indexInParent: number
//   arbType: number
//   arbSubType: number
// SenderNotPermitted = 13,
```

### CODE REVIEW

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (DZGoldman) [inline packages/arb-ts/src/lib/utils/arbProvider.ts:129]: Can we rename this param so it's clear that the caller of getBatch is responsible for providing an l1BlockNumber that corresponds to the receipt the sequencer gave for the tx with the provided seqnum?

(Maybe "l2TxnSeqNum" and "l2TxL1BlockNumber"

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (DZGoldman) [inline packages/arb-ts/src/lib/utils/arbProvider.ts:139]: I don't we need any delta here; i.e., the actual current l1 Block Number will always be greater than (or maybe equal to) the l1BlockNumber that the Sequencer assigns (since the Sequencer has a delayed view of the L1); the batch at which the transaction ultimately gets submitted will happen at an even greater l1 block number still; thus, I don't see any need to buffer negative/past direction.

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (DZGoldman) [inline packages/arb-ts/src/lib/utils/arbProvider.ts:147]: can use `Math.min`

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (DZGoldman) [inline packages/arb-ts/src/lib/utils/arbProvider.ts:144]: This upper bound only applies to transactions submitted in batches by the sequencer ( they indeed can't be included after `l1BlockNumber + delayBlocks`). Transactions that are force-included, however, can be force-included after any arbitrary amount of time. So since this `getBatch`method wants to cover `DelayedInboxForcedEvent`, as is, it would have to search past `l1BlockNumber + delayBlocks`.

Comment (DZGoldman) [reply packages/arb-ts/src/lib/utils/arbProvider.ts:144]: There's a few options here: one is to do the initial event log query only on `SequencerBatchDelivered` and `SequencerBatchDelivered` (using the `delayedBlockMax ` var as is), and then iff those return nil, _and_ the current L1 block is past the point of l1BlockNumber + delayBlocks, we do a second query for `DelayedInboxForced`. Unlovely to have to do 2 event queries, but  the 2nd query actually would only happen for a very unusual case.

**Comment by DZGoldman** [comment]
@yahgwai  Addresses question 1 in a comment; for 2 & 3, I'm not sure off the bat myself; play to look seems to be `/packages/arb-rpc-node/web3/eth.go`, tho it may be easiest to look in the node-eng team here.

(generally looking good — I have more to review still / want to make sure I didn't mess anything in the reorganisation diff, but left some comments)

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/utils/arbProvider.ts:129]: renamed, and added comment

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/utils/arbProvider.ts:139]: removed

File: packages/arb-ts/src/lib/utils/arbProvider.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/utils/arbProvider.ts]: Updated

**Comment by yahgwai** [comment]
I've some fields that we dont need to surface at the moment - we can investigate their nullability and document them properly later

**Review by DZGoldman — APPROVED** [review]
Little thing we can do is add the max_delay values to the network config (so those are basically constant / admin updated) to save on queries. Doesn't have to happen here tho; LGTM

File: packages/arb-ts/src/lib/dataEntities/arbTransaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:32]: I think this name is confusing, I'd instead call it FeeComponents, or ResourcesCharged, idk
But FeeStats and FeeStat right now is confusing

File: packages/arb-ts/src/lib/dataEntities/arbTransaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:167]: I believe this is incorrect. l1InboxBatchInfo is only returned if the L2 node explicitly enables it (which most don't and won't)

Comment (fredlacs) [reply packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:167]: context https://github.com/OffchainLabs/arbitrum/releases/tag/v1.0.0

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L1Transaction.ts:85]: hm I'd argue its cleaner if we change this to `getInboxMessages` and return the event.
Its pretty easy for the consumer to then do: `receipt.getInboxMessages().messageNum`

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L1Transaction.ts:90]: Is there a particular reason to listen to the inbox event here?
Given this is somewhat general I'd argue in favour for us to instead to check for the Bridge event.
It also includes the message number `messageIndex`
https://github.com/OffchainLabs/arbitrum/blob/005badd0f7beff05c222bf9d6d301564aa2b8f10/packages/arb-bridge-eth/contracts/bridge/interfaces/IBridge.sol#L22-L29

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L1Transaction.ts:146]: Is this needed?
It doesn't actually have a performance advantage compared to `getL1ToL2Messages` 
Just adds some message count validation then returns the desired index

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L1Transaction.ts:173]: I think we should make it more explicit that this refers to an asset deposit
```suggestion
  public getTokenDepositEvents(): DepositInitiatedEvent['args'][] {
```

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:85]: would be consistent with
`public getDepositEvents(): DepositInitiatedEvent['args'][]`

File: packages/arb-ts/src/lib/message/L2Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L2Transaction.ts:166]: I don't believe this is correct (mentioned in previous comment about `l1InboxBatchInfo`)
We need to instead do lookups against the L1 provider.
I agree `isDataAvailable` is the most precise way of describing this, but I don't think its the most intuitive for devs consuming the lib

File: packages/arb-ts/src/lib/message/L2Transaction.ts
Comment (fredlacs) [inline packages/arb-ts/src/lib/message/L2Transaction.ts:178]: I don't think it makes sense to wait for anything more than 1 confirmation.
I think we should throw an error if the user sets confirmations > 1

File: packages/arb-ts/src/lib/message/L2Transaction.ts
Comment (fredlacs) [reply packages/arb-ts/src/lib/message/L2Transaction.ts:166]: Will need to use the getBatch method from the arb provider

File: packages/arb-ts/src/lib/dataEntities/arbTransaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:32]: Updated to FeeStatComponents

File: packages/arb-ts/src/lib/dataEntities/arbTransaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:167]: I think it is correct, in getArbTransactionReceipt we use an l1 provider to fetch this info, if the provider is there.

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:85]: Updated

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:90]: yeah, defo, updated

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:146]: Yeah, I kinda agree. @DZGoldman mentioned that this is the most normal way of consuming the messages, and so was useful to have.

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:173]: Updated

File: packages/arb-ts/src/lib/message/L2Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L2Transaction.ts:166]: I kinda like the name - dyou have any other suggestions? In the comment we mention that this is referring specifically to the data being available on L1. It's also a more general principle rather than referring to batches.

File: packages/arb-ts/src/lib/message/L2Transaction.ts
Comment (yahgwai) [reply packages/arb-ts/src/lib/message/L2Transaction.ts:178]: Ignoring it for now

File: packages/arb-ts/src/lib/dataEntities/arbTransaction.ts
Comment (fredlacs) [reply packages/arb-ts/src/lib/dataEntities/arbTransaction.ts:167]: Ah I misinterpreted it, all good

File: packages/arb-ts/src/lib/message/L1Transaction.ts
Comment (fredlacs) [reply packages/arb-ts/src/lib/message/L1Transaction.ts:146]: We never use this internally, and externally its an extra couple of lines of validation which we let the consumer decide on

**Pull request merged**
